### ISSUE TITLE
Transferring encrypted data

### ISSUE THREAD

**Issue by zoolyka** [ricmoo/aes-js#13]
Hello!

It is not quite straightforward how to transfer the encrypted data, since its a byte array.

Many could think (as I did) that the encrypted byte array could be converted to a string with aesjs.util.convertBytesToString, being transferred over the wire, then converted back to a byte array with aesjs.util.convertStringToBytes for decryption.

Sadly this is not the case, since the included functions fails on these tasks (I guess the utilized encodeURI and such functions cannot handle the encrypted strings).

I've figured out that altering the above mentioned 2 functions to Google Closures "stringToByteArray" and "byteArrayToString" ones solves the issues, and its even faster.

(Hint link: https://github.com/google/closure-library/blob/e877b1eac410c0d842bcda118689759512e0e26f/closure/goog/crypt/crypt.js)

Sincerely,
Zoltan Hajdu

**Comment by zoolyka** [comment]
_Surely I'm talking about the utf8 string conversions, not the hex ones._

**Comment by ricmoo** [comment]
This has come up before, and I am considering removing the functions altogether, as their purpose was more to replace `new Buffer(someString, 'utf8')`. The important distinction is that encryption takes binary data and returns binary data, as well decryption takes binary data and returns binary data.

UTF-8 is not an encoding for binary data. Most(1) binary data will actually be completely invalid UTF-8 strings. To encode/decode binary data as a string, you need to use a binary-to-text encoding. But it is important to realize that it is quite likely that `decodeUtf8(encodeUtf8(something)) !== something`.

Please check out [this comment](https://github.com/ricmoo/aes-js/issues/11#issuecomment-205108162) for more details.

tl;dr - You can pass  second parameter into `aesjs.util.convertBytesToString(data, 'hex')` to use the hexadecimal binary encoding.

Notes: (1) Not technically *most*, as both the sets of binary data that are and are not valid UTF-8 strings are countable infinite (ie. Aleph null), but most was meant as hyperbole. :)

**Comment by ricmoo** [comment]
Also, looking at those google functions, I would highly recommend **not** using that as a method to serialize/deserialize binary data. If it ever passes through anything *string*y in JavaScript, it will get corrupted:

```javascript
// Notice 3 bytes in...
> var data = [0x75, 0xcc, 0x88];
undefined

// And their decimal values
> data
[ 117, 204, 136 ]

// And now notice 5 bytes out... (after using Google's code)
> new Buffer(String.fromCharCode.apply(null, data))
<Buffer 75 c3 8c c2 88>
```

**Comment by zoolyka** [comment]
Thank you for the detailed answer. Maybe extending these functions with binary data handling (aside of utf8 and hex) would be a nicer idea instead of dropping them :)

**Comment by ricmoo** [comment]
Yeah, this was mainly to avoid people having to include a Buffer implementation... I may rename it and add base64 in the future... Still on the fence. I may just add a section in the readme regarding binary encoding... Thanks! :)

**Comment by zoolyka** [comment]
Finally I've ended up with 5 functions which covers all my needs :)

1. Hex to bytes - use hex format keys for the encryption
2. String to bytes - prepare json encoded data (or any string) for encryption (which needs bytes)
3. Bytes to string - reassemble the json encoded data (or any string) after decryption (which produces bytes)
4. Base64 to bytes - prepare the base64 encoded data for decryption (which needs bytes) after its transferred over the wire
5. Bytes to base64 - convert the encrypted data (which produced bytes) to base64 format for transferring over the wire

`// Convert hex to bytes

function convertHexToBytes ( hex ) {

  var bytes = [];

  for (
    var i = 0,
        j = hex.length;
    i < j;
    i += 2
  ) {
    bytes.push(parseInt(hex.substr(i, 2), 16));
  };

  return(bytes);

};

// Convert string to bytes

function convertStringToBytes ( string ) {

  var bytes = [],
      p = 0;

  for (
    var i = 0;
    i < string.length;
    ++i
  ) {

    var c = string.charCodeAt(i);

    while (
      c > 0xff
    ) {
      bytes[p++] = ( c & 0xff );
      c >>= 8;
    };

    bytes[p++] = c;

  };

  return(bytes);

};

// Convert bytes to string

function convertBytesToString ( bytes ) {

  var chunk_size = 8192;

  if (
    bytes.length <= chunk_size
  ) {
    return(String.fromCharCode.apply(null, bytes));
  };

  var text = '';

  for (
    var i = 0;
    i < bytes.length;
    i += chunk_size
  ) {
    var chunk = bytes.slice(i, i + chunk_size);
    text += String.fromCharCode.apply(null, chunk);
  };

  return(text);

};

// Convert base64 to bytes

var Base64Encodings = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
var base64Lookup = [];

for (
  var i = 0;
  i < Base64Encodings.length;
  ++i
) {
  base64Lookup[Base64Encodings.charCodeAt(i)] = i;
};

Base64Encodings = Base64Encodings.split('');

function convertBase64ToBytes ( base64 ) {

  base64 = base64.replace(/[^A-Za-z0-9\+\/\=]+/g, '');

  var bytes = [],
      base64Length = base64.length,
      a, b, c, d;

  if (
    ( base64Length % 4 ) != 0
  ) {
    return(bytes);
  };

  for (
    var i = 0;
    i < base64Length;
    i += 4
  ) {

    a = base64Lookup[base64.charCodeAt(i)];
    b = base64Lookup[base64.charCodeAt(( i + 1 ))];
    c = base64Lookup[base64.charCodeAt(( i + 2 ))];
    d = base64Lookup[base64.charCodeAt(( i + 3 ))];

    bytes.push(( ( a << 2 ) | ( b >> 4 ) ));

    if (
      c != 64
    ) {
      bytes.push(( ( ( b & 15 ) << 4 ) | ( c >> 2 ) ));
    };

    if (
      d != 64
    ) {
      bytes.push(( ( ( c & 3 ) << 6 ) | ( d & 63 ) ));
    };

  };

  return(bytes);

};

// Convert bytes to base64

function convertBytesToBase64 ( bytes ) {

  var base64 = '',
      bytesLength = bytes.length,
      bytesRemainder = ( bytesLength % 3 );

  for (
    var i = 0;
    i < bytesLength;
    i += 3
  ) {
    base64 += Base64Encodings[( bytes[i] >> 2 )];
    base64 += Base64Encodings[( ( ( bytes[i] & 3 ) << 4 ) | ( bytes[( i + 1 )] >> 4 ) )];
    base64 += Base64Encodings[( ( ( bytes[( i + 1 )] & 15 ) << 2 ) | ( bytes[( i + 2 )] >> 6 ) )];
    base64 += Base64Encodings[( bytes[( i + 2 )] & 63 )];
  };

  if (
    bytesRemainder == 2
  ) {
    base64 = base64.substring(0, ( base64.length - 1 )) + '=';
  } else if (
    bytesRemainder == 1
  ) {
    base64 = base64.substring(0, ( base64.length - 2 )) + '==';
  };

  return(base64);

};`
